From 01eaef0d69a54e33b96b1fa9ff20c4e9ef678891 Mon Sep 17 00:00:00 2001
From: John Allen <john.allen@amd.com>
Date: Mon, 23 Sep 2024 15:41:12 +0000
Subject: [PATCH 07/14] x86/CPU/AMD: Only apply Zenbleed fix for Zen2 during
 late microcode load

A problem was introduced with f69759b ("x86/CPU/AMD: Move Zenbleed check to
the Zen2 init function") where a bit in the DE_CFG MSR is getting set after
a microcode late load.

The problem is that the microcode late load path calls into
amd_check_microcode and subsequently zen2_zenbleed_check. Since the patch
removes the cpu_has_amd_erratum check from zen2_zenbleed_check, this will
cause all non-Zen2 cpus to go through the function and set the bit in the
DE_CFG MSR.

Call into the zenbleed fix path on Zen2 cpus only.

Fixes: f69759be251d ("x86/CPU/AMD: Move Zenbleed check to the Zen2 init function")
Signed-off-by: John Allen <john.allen@amd.com>
Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 arch/x86/kernel/cpu/amd.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kernel/cpu/amd.c b/arch/x86/kernel/cpu/amd.c
index 1e0fe5f8a..65b2f0c5e 100644
--- a/arch/x86/kernel/cpu/amd.c
+++ b/arch/x86/kernel/cpu/amd.c
@@ -1218,5 +1218,6 @@ void amd_check_microcode(void)
 	if (boot_cpu_data.x86_vendor != X86_VENDOR_AMD)
 		return;
 
-	on_each_cpu(zenbleed_check_cpu, NULL, 1);
+	if (boot_cpu_has(X86_FEATURE_ZEN2))
+		on_each_cpu(zenbleed_check_cpu, NULL, 1);
 }
-- 
2.47.0

